#+TITLE: pipette
#+STARTUP: content
* Development
** TODO Can we process the new ssGSEA data?
** TODO Can we load the transcript-level data?
** TODO Consider adding MD5 checksum checks for DepMap files on Figshare.
** TODO Integrate mutation information from OncoKB.
** TODO Consider supporting z score calculations using reference population without copy number amplifications, similar to cBioPortal approach.
    https://github.com/cBioPortal/cbioportal/blob/master/docs/Z-Score-normalization-script.md
    https://github.com/cBioPortal/cbioportal/blob/master/docs/File-Formats.md#expression-data
** TODO plotHeatmap needs to support `cellLineName` labeling.
** TODO Need to update code for DEMETER2 processing.
** TODO Improve the CLI messages about what files we're downloading, not just the vague figshare URLs.
** TODO Simplify this to build everything into a single MultiAssayExpermient object.
    Refer to cBioPortalData package for inspiration.
** TODO Describe the colData columns using elementMetadata -- refer to DESeq2 for example.
** TODO Add support for Gygi lab proteomics and new Sanger proteomics dataset.
** TODO Convert YAML file to JSON in a future update.
** TODO Work on incorporating CCLE_segment_cn.csv into analysis.
** TODO Add option to download predictability data
    https://depmap.org/portal/gene/predictability_files
** TODO `Codependencies`: Consider adding parameterized support for multiple genes into `geneName1`.
    This may be useful for quickly testing multiple genes of interest per condition.
** TODO Consider splitting analyses into primary and metastatic disease subtypes.
** TODO CCLE mutation columns are a bit weird in camelCase.
    TCGA hotspot column needs to be fixed.
** TODO Add support for labeling specific cells in plots.
** TODO Add support for repurposing dataset
    https://depmap.org/repurposing/
