#+TITLE: DepMapAnalysis
#+STARTUP: content
* Development
** TODO Add support for downloading BAM files from DepMap CCLE AWS bucket.** TODO Add support for downloading BAM files from DepMap CCLE AWS bucket.** TODO Add support for downloading BAM files from DepMap CCLE AWS bucket.** TODO Add support for downloading BAM files from DepMap CCLE AWS bucket.** TODO Add support for downloading BAM files from DepMap CCLE AWS bucket.** TODO Add support for downloading BAM files from DepMap CCLE AWS bucket.** TODO Add support for downloading BAM files from DepMap CCLE AWS bucket.** TODO Add support for downloading BAM files from DepMap CCLE AWS bucket.
** TODO clinvarToOncokb annotation mappings
** TODO Add plotCleveland support for predictSensitivity return.
** TODO Add a quick lookup function to determine if a driver mutation is present in specific cell lines.
** TODO Add plotting function to visualize correlation of two gene effects.
** TODO Consider classing the `predictSensitivty` return so we can plug that into plots.
** TODO Ensure that "cellosaurus" slotted into colData remains classed as S4 Cellosaurus.
    This makes downstream method support for `selectCells` simpler.
** TODO Only genes that map to GRCh37 should be considered for biomarker analysis.
    Many clinical platforms only map to this reference genome.
** TODO Can we add OncoKB integration to the DepMapMutations class?
    https://github.com/waldronlab/oncoKBData/blob/devel/R/oncoKB.R
    https://github.com/oncokb/oncokb-annotator
    https://github.com/cBioPortal/cbioportal/blob/master/docs/Import-OncoKB-annotations-as-Custom-Driver-Annotations.md
    https://github.com/taylor-lab/annotateMaf
** TODO Migrate legacy CellModelPassports code here.
    In particular, we want to be able to easily access driver mutation info.
    Maybe we call this `DepMapDriverMutations`?
** TODO Consider renaming `DepMapMutation` to `DepMapMutations`?
** TODO Integrate mutation information from OncoKB.
    Look at the approach currently used on cBioPortal website.
** TODO Add option to download predictability data
    https://depmap.org/portal/gene/predictability_files
** TODO DepMapProteomics: Consider dropping the Nusinow et al. 2020 dataset in favor of just Goncalves et al 2022.
** TODO Consider dropping `DepMapMicroRNA` from the package.
** TODO Can we process the new ssGSEA data?
** TODO Can we load the transcript-level data?
    Should the transcript-level data go into `DepMapExpression` or a different class?
    OmicsExpressionProteinCodingGenesTPMLogp1.csv (genes)
    OmicsExpressionTranscriptsTPMLogp1Profile.csv (transcripts)
** TODO Consider adding MD5 checksum checks for DepMap files on Figshare.
** TODO Consider supporting z score calculations using reference population without copy number amplifications, similar to cBioPortal approach.
    https://github.com/cBioPortal/cbioportal/blob/master/docs/Z-Score-normalization-script.md
    https://github.com/cBioPortal/cbioportal/blob/master/docs/File-Formats.md#expression-data
** TODO plotHeatmap needs to support `cellLineName` labeling.
** TODO Need to update code for DEMETER2 processing.
** TODO Improve the CLI messages about what files we're downloading, not just the vague figshare URLs.
** TODO Simplify this to build everything into a single MultiAssayExpermient object.
    Refer to cBioPortalData package for inspiration.
** TODO Describe the colData columns using elementMetadata -- refer to DESeq2 for example.
** TODO Add support for Gygi lab proteomics and new Sanger proteomics dataset.
** TODO Convert YAML file to JSON in a future update.
** TODO Work on incorporating CCLE_segment_cn.csv into analysis.
** TODO `Codependencies`: Consider adding parameterized support for multiple genes into `geneName1`.
    This may be useful for quickly testing multiple genes of interest per condition.
** TODO Consider splitting analyses into primary and metastatic disease subtypes.
** TODO CCLE mutation columns are a bit weird in camelCase.
    TCGA hotspot column needs to be fixed.
** TODO Add support for labeling specific cells in plots.
** TODO Add support for repurposing dataset
    https://depmap.org/repurposing/
