## FIXME Assert that no classes that extend SE contain any cellLineName
## with NA in colData.

## FIXME Need to deal with this: anyNA(assay(crispr))



#' Sample metadata column names, defined in colData
#'
#' @note Updated 2022-09-21.
#' @noRd
.expectedColData <- c(
    "age",
    "alias",
    "ccleName",
    "cellLineName",
    "cellosaurusId",
    "cellosaurusIssues",
    "cosmicId",
    "defaultGrowthPattern",
    "depmapId",
    "depmapPublicComments",
    "lineage",
    "lineageMolecularSubtype",
    "lineageSubSubtype",
    "lineageSubtype",
    "modelManipulation",
    "modelManipulationDetails",
    "ncitDiseaseId",
    "ncitDiseaseName",
    "parentDepmapId",
    "patientId",
    "primaryDisease",
    "primaryOrMetastasis",
    "sampleCollectionSite",
    "sangerModelId",
    "sex",
    "source",
    "strippedCellLineName",
    "subtype",
    "wtsiMasterCellId"
)



#' Expected DataFrame metadata
#'
#' @note Updated 2022-03-09.
#' @noRd
.expectedDFMetadata <- list(
    "dataset" = "character",
    "date" = "Date",
    "packageName" = "character",
    "packageVersion" = "package_version"
)



#' Expected SummarizedExperiment metadata
#'
#' @note Updated 2022-09-07.
#' @noRd
.expectedSEMetadata <- append(
    x = .expectedDFMetadata,
    values = list(
        ## JSON was added in 0.3.0 (2022-09).
        ## > "json" = "list",
        ## > "missingCells" = "character",
        ## > "retiredGenes" = "character",
        ## > "wd" = "character",
        ## YAML was removed in favor of JSON in 0.3.0.
        ## > "yaml" = "list",
        "sessionInfo" = "sessionInfo"
    )
)



#' CCLE copy number data
#'
#' @details
#' Gene level copy number data, log2 transformed with a pseudo count of 1.
#' This is generated by mapping genes onto the segment level calls.
#'
#' Inherits from `SummarizedExperiment`.
#' Cells in columns, genes in rows.
#'
#' @export
#' @note Updated 2022-08-19.
#'
#' @return `CCLECopyNumberData`.
setClass(
    Class = "CCLECopyNumberData",
    contains = "SummarizedExperiment"
)
setValidity(
    Class = "CCLECopyNumberData",
    method = function(object) {
        ok <- validate(
            hasRownames(object),
            allAreMatchingRegex(
                x = rownames(object),
                pattern = "^[_A-Za-z0-9]+_[0-9]+$"
            ),
            hasColnames(object),
            allAreMatchingRegex(
                x = colnames(object),
                pattern = "^ACH_[0-9]{6}$"
            ),
            isSubset("log2CopyNumber", assayNames(object)),
            isSubset(
                x = c(
                    ## > "xTaxId"
                    "chromosome",
                    "dbXrefs",
                    "description",
                    "featureType",
                    "geneId",
                    "geneName",
                    "geneSynonyms",
                    "mapLocation",
                    "modificationDate",
                    "nomenclatureStatus",
                    "otherDesignations",
                    "typeOfGene"
                ),
                y = colnames(rowData(object))
            ),
            isSubset(
                x = .expectedColData,
                y = colnames(colData(object))
            )
        )
        if (!isTRUE(ok)) {
            return(ok)
        }
        ok <- validateClasses(
            object = metadata(object),
            expected = .expectedSEMetadata,
            subset = TRUE
        )
        if (!isTRUE(ok)) {
            return(ok)
        }
        TRUE
    }
)



#' CCLE expression data
#'
#' @details
#'
#' RNA-seq TPM gene expression data for just protein coding genes using RSEM.
#' Log2 transformed, using a pseudo-count of 1.
#'
#' Inherits from `SummarizedExperiment`.
#' Cells in columns, genes in rows.
#'
#' @export
#' @note Updated 2022-08-19.
#'
#' @return `CCLEExpressionData`.
setClass(
    Class = "CCLEExpressionData",
    contains = "SummarizedExperiment"
)
setValidity(
    Class = "CCLEExpressionData",
    method = function(object) {
        ok <- validate(
            hasRownames(object),
            allAreMatchingRegex(
                x = rownames(object),
                pattern = "^[_A-Za-z0-9]+_[0-9]+$"
            ),
            hasColnames(object),
            allAreMatchingRegex(
                x = colnames(object),
                pattern = "^ACH_[0-9]{6}$"
            ),
            isSubset("log2Tpm", assayNames(object)),
            isSubset(
                x = c(
                    ## > "xTaxId"
                    "chromosome",
                    "dbXrefs",
                    "description",
                    "featureType",
                    "geneId",
                    "geneName",
                    "geneSynonyms",
                    "mapLocation",
                    "modificationDate",
                    "nomenclatureStatus",
                    "otherDesignations",
                    "typeOfGene"
                ),
                y = colnames(rowData(object))
            ),
            isSubset(
                x = .expectedColData,
                y = colnames(colData(object))
            )
        )
        if (!isTRUE(ok)) {
            return(ok)
        }
        ok <- validateClasses(
            object = metadata(object),
            expected = .expectedSEMetadata,
            subset = TRUE
        )
        if (!isTRUE(ok)) {
            return(ok)
        }
        TRUE
    }
)



#' CCLE fusion data
#'
#' @details
#' Inherits from `DFrame`.
#'
#' @note Updated 2022-09-21.
#' @export
#'
#' @return `CCLEFusionData`.
setClass(
    Class = "CCLEFusionData",
    contains = "DFrame"
)
setValidity(
    Class = "CCLEFusionData",
    method = function(object) {
        ok <- validate(
            isSubset(
                x = c(
                    "annots",
                    "ccleCount",
                    "depmapId",
                    "ffpm",
                    "fusionName",
                    "junctionReadCount",
                    "largeAnchorSupport",
                    "leftBreakDinuc",
                    "leftBreakEntropy",
                    "leftBreakpoint",
                    "leftGene",
                    "rightBreakDinuc",
                    "rightBreakEntropy",
                    "rightBreakpoint",
                    "rightGene",
                    "spanningFragCount",
                    "spliceType"
                ),
                y = colnames(object)
            )
        )
        if (!isTRUE(ok)) {
            return(ok)
        }
        ok <- validateClasses(
            object = metadata(object),
            expected = append(
                x = .expectedDFMetadata,
                values = list(
                    "filtered" = "logical"
                )
            ),
            subset = TRUE
        )
        if (!isTRUE(ok)) {
            return(ok)
        }
        TRUE
    }
)



#' CCLE microRNA expression data
#'
#' @details
#' Inherits from `SummarizedExperiment`.
#' Cells in columns, microRNAs in rows.
#'
#' Same as the re-released `CCLE_miRNA_20181103.gct` file.
#'
#' @export
#' @note Updated 2023-01-25.
#'
#' @return `CCLEMicroRNAExpressionData`.
setClass(
    Class = "CCLEMicroRNAExpressionData",
    contains = "RangedSummarizedExperiment"
)
setValidity(
    Class = "CCLEMicroRNAExpressionData",
    method = function(object) {
        ## FIXME Need to tighten this up.
        TRUE
    }
)



#' CCLE mutation data
#'
#' @details
#' Inherits from `DFrame`.
#'
#' @note Updated 2022-09-21.
#' @export
#'
#' @return `CCLEMutationData`.
setClass(
    Class = "CCLEMutationData",
    contains = "DFrame"
)
setValidity(
    Class = "CCLEMutationData",
    method = function(object) {
        ok <- validate(
            isSubset(
                x = c(
                    "alternateAllele",
                    "annotationTranscript",
                    "cdnaChange",
                    "cgaWesAc",
                    "chromosome",
                    "codonChange",
                    "cosmicHsCnt",
                    "dbsnpRs",
                    "dbsnpValStatus",
                    "depmapId",
                    "endPosition",
                    "entrezGeneId",
                    "exacAf",
                    "geneName",
                    "genomeChange",
                    "hcAc",
                    "isCosmicHotspot",
                    "isDeleterious",
                    "isTcgaHotspot",
                    "ncbiBuild",
                    "proteinChange",
                    "rdAc",
                    "referenceAllele",
                    "rnaseqAc",
                    "sangerWesAc",
                    "startPosition",
                    "strand",
                    "tcgaHsCnt",
                    "variantAnnotation",
                    "variantClassification",
                    "variantType",
                    "wgsAc"
                ),
                y = colnames(object)
            )
        )
        if (!isTRUE(ok)) {
            return(ok)
        }
        ok <- validateClasses(
            object = metadata(object),
            expected = .expectedDFMetadata,
            subset = TRUE
        )
        if (!isTRUE(ok)) {
            return(ok)
        }
        TRUE
    }
)



#' Gene effect co-dependencies
#'
#' @details
#' Inherits from `DataFrame`.
#'
#' @note Updated 2022-08-05.
#' @export
#'
#' @return `Codependencies`.
setClass(
    Class = "Codependencies",
    contains = "DFrame"
)
setValidity(
    Class = "Codependencies",
    method = function(object) {
        ok <- validate(
            identical(
                x = c(
                    "geneName1",
                    "geneName2",
                    "pearson"
                ),
                y = colnames(object)
            )
        )
        if (!isTRUE(ok)) {
            return(ok)
        }
        ok <- validateClasses(
            object = metadata(object),
            expected = .expectedDFMetadata,
            subset = TRUE
        )
        if (!isTRUE(ok)) {
            return(ok)
        }
        TRUE
    }
)



#' Gene effect in cancer cell lines
#'
#' @details
#' Inherits from `SummarizedExperiment`.
#' Cells in columns, genes in rows.
#'
#' @export
#' @note Updated 2022-08-19.
#'
#' @return `GeneEffect`.
#'
#' @seealso
#' - https://depmap.org/portal/achilles/
#' - https://depmap.org/ceres/
#' - https://score.depmap.sanger.ac.uk/
setClass(
    Class = "GeneEffect",
    contains = "SummarizedExperiment"
)
setValidity(
    Class = "GeneEffect",
    method = function(object) {
        ok <- validate(
            hasRownames(object),
            allAreMatchingRegex(
                x = rownames(object),
                pattern = "^[_A-Za-z0-9]+_[0-9]+$"
            ),
            hasColnames(object),
            allAreMatchingRegex(
                x = colnames(object),
                pattern = "^ACH_[0-9]{6}$"
            ),
            isSubset("effect", assayNames(object))
        )
        if (!isTRUE(ok)) {
            return(ok)
        }
        ok <- validateClasses(
            object = metadata(object),
            expected = append(
                x = .expectedSEMetadata,
                values = list(
                    "libraryType" = "character",
                    "scoringMethod" = "character"
                )
            ),
            subset = TRUE
        )
        if (!isTRUE(ok)) {
            return(ok)
        }
        switch(
            EXPR = metadata(object)[["libraryType"]],
            "crispr" = {
                ok <- validate(
                    isSubset("probability", assayNames(object)),
                    isSubset(
                        x = .expectedColData,
                        y = colnames(colData(object))
                    )
                )
                if (!isTRUE(ok)) {
                    return(ok)
                }
                ok <- validateClasses(
                    object = metadata(object),
                    expected = list(
                        "commonEssentials" = "character",
                        "controlCommonEssentials" = "character",
                        "controlNonessentials" = "character"
                    ),
                    subset = TRUE
                )
                if (!isTRUE(ok)) {
                    return(ok)
                }
            },
            "rnai" = {
                ok <- validate(
                    isSubset(
                        x = .expectedColData,
                        y = colnames(colData(object))
                    ),
                    isSubset(
                        x = c(
                            "inAchilles",
                            "inDrive",
                            "inMarcotte",
                            "marcotteName",
                            "marcotteSubtypeIntrinsic",
                            "marcotteSubtypeNeve",
                            "marcotteSubtypeThreeReceptor",
                            "novartisName",
                            "novartisPathologistAnnotation",
                            "novartisPrimarySite"
                        ),
                        y = colnames(colData(object))
                    )
                )
                if (!isTRUE(ok)) {
                    return(ok)
                }
            }
        )
        TRUE
    }
)
